File: DSA-C++/6-array/12-single-number.cpp
#include <bits/stdc++.h>

using namespace std;
// Given a non-empty array of integers nums, every element appears twice except for one. Find that single one.


// brute force - O(n²), O(1)
int single_number_1(vector <int> &nums) {
    int n = nums.size();

    for (int i=0; i<n; ++i) {
        int cnt = 0;
        for (int j=0; j<n; ++j) {
            if (nums[i] == nums[j]) {
                cnt++;
            }
        }

        if (cnt < 2) {
            return nums[i];
        }
    }

    return -1;
}


// hashing -O(n), O(n)
int single_number_2(vector <int> &nums) {
    int n = nums.size();
    unordered_map <int, int> umap;

    for (int &x:nums) {
        umap[x]++;
    }

    for (auto [k,v]:umap) {
        if (v < 2) {
            return k;
        }
    }

    return -1;
}


// xor elements - O(n), O(1)
int single_number_3(vector <int> &nums) {
    int xored=0;

    for (int &x:nums) {
        xored ^= x;
    }

    return xored;
}

int main() {
    vector <int> nums = {1,2,1,2,4};
    // nums = {1,2,4,1,2};
    // nums = {4,1,2,1,2};

    // cout << single_number_1(nums) << endl;

    // cout << single_number_2(nums) << endl;

    cout << single_number_3(nums) << endl;


    return 0;
}

File: DSA-Python/5-array/12-single-number.py
# Given a non-empty array of integers nums, every element appears twice except for one. Find that single one.



# brute force - O(n²), O(1)
def single_number_1(nums:list[int]) -> int:
    n = len(nums)
    
    for i in range(n):
        cnt = 0
        for j in range(n):
            if nums[i] == nums[j]:
                cnt += 1
                
        if cnt < 2:
            return nums[i]
        
    return -1



# hashing - O(n), O(n)
def single_number_2(nums: list[int]) -> int:
    umap = {}
    
    for x in nums:
        umap[x] = umap.get(x, 0) + 1
        
    for (k,v) in umap.items():
        if v < 2:
            return k
    
    return 1
    
    

# xor approach - O(n), O(1)
def single_number_3(nums: list[int]) -> int:
    xor = 0
    
    for x in nums:
        xor ^= x
        
    return xor



if __name__ == "__main__": 
    loop = int(input())

    for _ in range(loop):
        nums = list(map(int, input().split(' ')))
        
        res = single_number_1(nums)
        
        res = single_number_2(nums)
        
        res = single_number_3(nums)
        
        print(res)